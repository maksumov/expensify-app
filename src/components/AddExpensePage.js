import React from "react";

const AddExpensePage = () => <div>This is from my AddExpensePage</div>;

export default AddExpensePage;
