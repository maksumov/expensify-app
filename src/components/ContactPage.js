import React from "react";

const ContactPage = () => (
  <div>
    <h2>Contact me</h2>
    <p>You can contact me at maxumov@gmail.com</p>
  </div>
);

export default ContactPage;
