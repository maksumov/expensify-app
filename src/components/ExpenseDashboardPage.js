import React from "react";

const ExpenseDashboardPage = () => (
  <div>This is from my ExpenseDashboardPage</div>
);

export default ExpenseDashboardPage;
