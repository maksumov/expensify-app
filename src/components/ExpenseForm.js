import React, { Component } from 'react'
import moment from 'moment'
import 'react-dates/initialize'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'

// moment.locale('ru')
// const now = moment()
// console.log(now.format('DD.MM.YYYY'))

export default class ExpenseForm extends Component {
  state = {
    description: '',
    note: '',
    amount: '',
    createdAt: moment(),
    focused: false,
    error: '',
  }
  onDescriptionChange = (e) => {
    const description = e.target.value

    // Clear error state
    // if (this.state.error && description && this.state.amount) {
    //   this.setState(() => ({ error: '' }))
    // }
    this.setState(() => ({ description }))
  }
  onNoteChange = (e) => {
    const note = e.target.value
    this.setState(() => ({ note }))
  }
  onAmountChange = (e) => {
    const amount = e.target.value

    if (amount.match(/^\d*(\d\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))

      // Clear error state
      // if (this.state.error && amount && this.state.description) {
      //   this.setState(() => ({ error: '' }))
      // }
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      console.log(createdAt)
      this.setState(() => ({ createdAt }))
    }
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ focused }))
  }
  onSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      // Set error state equal to 'Please provide description and amount.'
      this.setState(() => ({ error: 'Please provide description and amount.' }))
    } else {
      // clear the error
      this.setState(() => ({ error: '' }))

      this.props.onSubmit({
        description: this.state.description,
        amount: Math.round(parseFloat(this.state.amount, 10) * 100),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      })
      console.log('submitted!')
    }
  }
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type='text'
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input type='text' placeholder='Amount' value={this.state.amount} onChange={this.onAmountChange} />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            id='datepicker'
            displayFormat='DD.MM.YYYY'
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder='Add a note for your expense (optional)'
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add expense</button>
        </form>
      </div>
    )
  }
}
