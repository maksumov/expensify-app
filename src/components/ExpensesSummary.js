import React from 'react'
import { connect } from 'react-redux'
import expensesTotal from '../selectors/expenses-total'
import expensesCount from '../selectors/expenses-count'
import selectExpenses from '../selectors/expenses'
import numeral from 'numeral'

export const ExpensesSummary = (props) => {
  const total = expensesTotal(props.expenses)
  const count = expensesCount(props.expenses)

  return (
    <div>
      {count > 0 &&
        (count === 1 ? (
          <p>
            Viewing {count} expense totalling {numeral(total / 100).format('$0,0.00')}
          </p>
        ) : (
          <p>
            Viewing {count} expenses totalling {numeral(total / 100).format('$0,0.00')}
          </p>
        ))}
    </div>
  )
}

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  }
}

export default connect(mapStateToProps)(ExpensesSummary)
