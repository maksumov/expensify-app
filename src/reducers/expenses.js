// Expenses Reducer
const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense]
    case 'REMOVE_EXPENSE':
      return state.filter((item) => item.id !== action.id)
    case 'EDIT_EXPENSE':
      return state.map((item) => {
        if (item.id !== action.id) return item
        // else return Object.assign({}, item, action.updates);
        else return { ...item, ...action.updates }
      })
    case 'SET_EXPENSES':
      return action.expenses
    default:
      return state
  }
}
