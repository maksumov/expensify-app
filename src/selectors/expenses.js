import moment from 'moment'

// Get Visible Expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt)
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
      const textMatch = !text || expense.description.toLowerCase().indexOf(text.toLowerCase()) !== -1

      // console.log('startDateMatch', startDateMatch, startDate, expense.createdAt)
      // console.log('endDateMatch', endDateMatch, endDate, expense.createdAt)
      // console.log('textMatch', textMatch, text, expense.description)
      return startDateMatch && endDateMatch && textMatch
    })
    .sort((a, b) => {
      // Sorting by date (recent first)
      if (sortBy === 'date') {
        return a.createdAt > b.createdAt ? -1 : 1
      }
      // Sorting by amount (biggest first)
      else if (sortBy === 'amount') {
        return b.amount - a.amount
      }
    })
}
