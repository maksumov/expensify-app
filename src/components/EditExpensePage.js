import React, { Component } from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpensePage extends Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense)
    this.props.history.push('/')
  }
  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id })
    this.props.history.push('/')
  }
  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove</button>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, updates) => dispatch(startEditExpense(id, updates)),
  startRemoveExpense: ({ id }) => dispatch(startRemoveExpense({ id })),
})

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id),
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
