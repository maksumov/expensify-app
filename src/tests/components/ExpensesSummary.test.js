import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'

test('should correctly render if no expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[]} />)
  expect(wrapper).toMatchSnapshot()
})

test('should correctly render with one expense', () => {
  const wrapper = shallow(<ExpensesSummary expenses={[expenses[1]]} />)
  expect(wrapper).toMatchSnapshot()
})

test('should correctly render with many expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenses={expenses} />)
  expect(wrapper).toMatchSnapshot()
})
