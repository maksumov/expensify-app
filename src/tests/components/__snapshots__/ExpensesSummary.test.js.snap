// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render if no expenses 1`] = `<div />`;

exports[`should correctly render with many expenses 1`] = `
<div>
  <p>
    Viewing 
    3
     expenses totalling 
    $1,141.95
  </p>
</div>
`;

exports[`should correctly render with one expense 1`] = `
<div>
  <p>
    Viewing 
    1
     expense totalling 
    $1,095.00
  </p>
</div>
`;
